src/riak_core_job_manager.erl

```diff
@@ -0,0 +1,1310 @@
+%% and relatively obvious.
+%%
+submit(Job) ->
+    case riak_core_job:version(Job) of
+        {job, _} ->
+            gen_server:call(?JOBS_MGR_NAME, {submit, Job});
+        _ ->
+            erlang:error(badarg, [Job])
+    end.
+
+%% ===================================================================
+%% Work Callback API
+%% ===================================================================
+
+-spec running_job(MgrKey :: mgr_key()) -> riak_core_job:job() | {error, term()}.
+%%
+%% @doc Returns the running Job associated with MgrKey.
+%%
+%% This function should only be invoked by a UoW while it's running.
+%% In any other case, it returns `{error, not_running}'.
+%%
+running_job(#mgrkey{mgr = Mgr, key = Key}) ->
+    gen_server:call(Mgr, {running_job, Key}).
+
+%% ===================================================================
+%% Runner Callback API
+%% ===================================================================
+
+-spec cleanup(MgrKey :: mgr_key()) -> ok.
+%% @private
+%% @doc Callback for the job runner indicating the UOW is being cleaned up.
+%%
+cleanup(MgrKey) ->
+    update_job(MgrKey, cleanup).
+
+-spec finished(MgrKey :: mgr_key(), Result :: term()) -> ok.
+%% @private
+%% @doc Callback for the job runner indicating the UOW is finished.
+%%
+finished(MgrKey, Result) ->
+    update_job(MgrKey, finished, Result).
+
+-spec running(MgrKey :: mgr_key()) -> ok.
+%% @private
+%% @doc Callback for the job runner indicating the UOW is being run.
+%%
+running(MgrKey) ->
+    update_job(MgrKey, running).
+
+-spec starting(MgrKey :: mgr_key()) -> ok.
+%% @private
+%% @doc Callback for the job runner indicating the UOW is being started.
+%%
+starting(MgrKey) ->
+    update_job(MgrKey, started).
+
+%% ===================================================================
+%% Gen_server API
+%% ===================================================================
+
+-spec code_change(OldVsn :: term(), State :: state(), Extra :: term())
+        -> {ok, state()}.
+%% @private
+% we don't care, just carry on
+%
+code_change(_, State, _) ->
+    {ok, State}.
+
+-spec handle_call(Msg :: term(), From :: {pid(), term()}, State :: state())
+        -> {reply, term(), state()} | {stop, term(), term(), state()}.
+%% @private
+%% @end
+%
+% config() -> config() | {error, term()}.
+%
+handle_call(config, _, State) ->
+    {reply, [
+        {?JOB_SVC_CONCUR_LIMIT, State#state.rmax},
+        {?JOB_SVC_HIST_LIMIT,   State#state.hmax},
+        {?JOB_SVC_QUEUE_LIMIT,  State#state.qmax}
+    ] ++ gen_server:call(?JOBS_SVC_NAME, config), State};
+%
+% cancel(JobId :: jid(), Kill :: boolean())
+%   ->  {ok, killed | canceled | history, job()}
+%       | {error, running, job()} | false | {error, term()}
+%
+handle_call({cancel, JobId, Kill}, _, State) ->
+    case c_find(JobId, State#state.loc) of
+
+        {_, Ref} when erlang:is_reference(Ref) ->
+            case Kill of
```

Comment:
Too many long, nested case statements here. At minimum, we should break this section out into something like a `cancel_by_reference` helper function with separate clauses for `Kill` being `true` or not.